File: backend/tests/utils/datasource.py
"""Test helpers for DataSource."""

from sqlmodel import Session

from app.core.config import settings
from app.models_dbapi import DataSource, ProductTypeEnum
from tests.utils.utils import random_lower_string


def create_random_datasource(
    db: Session,
    *,
    name: str | None = None,
    product_type: ProductTypeEnum = ProductTypeEnum.POSTGRES,
    host: str | None = None,
    port: int | None = None,
    database: str | None = None,
    username: str | None = None,
    password: str | None = None,
    is_active: bool = True,
) -> DataSource:
    """Create a DataSource in the DB. Uses app Postgres settings by default so test/test and preTest can connect."""
    ds = DataSource(
        name=name or f"ds-{random_lower_string()}",
        product_type=product_type,
        host=host or settings.POSTGRES_SERVER,
        port=port or settings.POSTGRES_PORT,
        database=database or settings.POSTGRES_DB,
        username=username or settings.POSTGRES_USER,
        password=password or settings.POSTGRES_PASSWORD,
        is_active=is_active,
    )
    db.add(ds)
    db.commit()
    db.refresh(ds)
    return ds

File: frontend/tests/utils/datasource.ts
import { api } from "./apiClient.ts"

export type DataSourceRecord = {
  id: string
  name: string
  product_type: string
  host: string
  port: number
  database: string
  username: string
  is_active: boolean
}

export type CreateDataSourceInput = {
  name: string
  product_type?: "postgres" | "mysql" | "trino" | "minio"
  host?: string
  port?: number
  database?: string
  username?: string
  password?: string
  description?: string | null
}

export async function createDataSource(
  input: CreateDataSourceInput,
): Promise<DataSourceRecord> {
  return api.post<DataSourceRecord>("/api/v1/datasources/create", {
    name: input.name,
    product_type: input.product_type ?? "postgres",
    host: input.host ?? "localhost",
    port: input.port ?? 5432,
    database: input.database ?? "app",
    username: input.username ?? "postgres",
    // Default to the real Postgres password from .env so the SQL engine can
    // actually connect; individual tests can still override.
    password: input.password ?? process.env.POSTGRES_PASSWORD ?? "postgres",
    description: input.description ?? null,
    is_active: true,
  })
}

export async function deleteDataSource(id: string): Promise<void> {
  await api.delete(`/api/v1/datasources/delete/${id}`)
}

export async function deleteDataSourcesMatching(prefix: string): Promise<void> {
  const res = await api.post<{ data: Array<{ id: string; name: string }> }>(
    "/api/v1/datasources/list",
    { page: 1, page_size: 100 },
  )
  await Promise.all(
    res.data
      .filter((d) => d.name.startsWith(prefix))
      .map((d) => deleteDataSource(d.id).catch(() => {})),
  )
}
